torch/distributed/fsdp/_common_utils.py

```diff
@@ -184,6 +201,47 @@ def _is_fsdp_flattened(tensor: torch.Tensor) -> bool:
     return getattr(tensor, FSDP_FLATTENED, False)
 
 
+def _get_handles_to_fqns(
+    model: torch.nn.Module,
+) -> Dict[nn.Parameter, List[str]]:
+    """
+    It is used for composable fully_shard() code path, constructs a mapping from
+    flat_param in the FlatParamHandle to a list of its FQNs. Each
+    ``FlatParameter`` maps to a list of its original parameter FQNs, which may
+    have length greater than one. All FQNs are prefixed starting from
+    ``model``.
+
+    Args:
+        model (torch.nn.Module): Root module (which may or may not be passed to
+                                 composable `fully_shard()`).
+    """
+
+    def module_fn(module, prefix, param_to_fqns):
+        state = _get_module_fsdp_state(module)
+        if state is None:
+            return
+        handles = state._fully_sharded_module_to_handles.get(module, [])
+        for handle in handles:
+            param = handle.flat_param
+            assert type(param) is flat_param_file.FlatParameter
+            global_fqns = [
+                clean_tensor_name(prefix + name) for name in param._fqns
+            ]  # prefixed from the top level `model` (i.e. including `prefix`)
+            param_to_fqns[param] = global_fqns
+
+    def return_fn(param_to_fqns):
+        return param_to_fqns
+
+    param_to_unflat_param_names: Dict[torch.nn.Parameter, List[str]] = {}
```

Comment:
nit: I feel like we have some inconsistency in the naming that we should try to cut down on now. Originally, when I first onboarded, I had a mental model of "flat" parameter names (actually `FlatParameter` FQNs) and "unflat" parameter names (actually original parameter FQNs), which is why so much of the optimizer state checkpointing code uses the terms like `unflat_param_names`.

However, since then, I learned you guys use the term FQN, so I started migrating most of the core code to do that, only leaving optimizer state checkpointing not-yet-migrated.

In this case, I feel we should change `return_fn` to use `flat_param_to_fqns` instead of `param_to_fqns`. If there are variable shadowing issues, then you can prefix with `_` for `return_fn()`.

Then, maybe we can change `param_to_unflat_param_names` to also be `flat_param_to_fqns`. This makes it clearer throughout that it is the same quantity (being populated in the `module_fn`, being returned from this helper, and being used downstream in the unit test).